:bug: Add toString support to remote functions

twolfson opened the pull request:
In https://github.com/twolfson/karma-electron/issues/20, we received a bug report of not being able to stringify remote `electron` functions. After some searching, we found this was being caused by `new Proxy` which doesn't support `toString()` in any scenario:

https://github.com/electron/electron/blob/v1.6.3/lib/renderer/api/remote.js#L170-L190

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/toString

```js
new Proxy(function x () {}, {}).toString()
// VM594:1 Uncaught TypeError: Function.prototype.toString is not generic
//    at Proxy.toString (<anonymous>)
//    at <anonymous>:1:33
```

We have implemented a fallback implementation for Karma (where this issue was occurring):

https://github.com/karma-runner/karma/pull/2595

but believe we should fix the source of the problem by providing a `toString` proxy solution. In this PR:

- Added bound target `toString` function when `toString` is requested via `remote` proxy
- Added test

The result looks like:

```js
electron.remote.clipboard.write.toString();
// function (...args) {
//   if (this && this.constructor === remoteMemberFunction) {
//     // Constructor call.
//     let ret = ipcRenderer.sendSync('ELECTRON_BROWSER_MEMBER_CONSTRUCTOR', metaId, member.name, wrapArgs(args))
//     return metaToValue(ret)
//   } else {
//     // Call member function.
//     let ret = ipcRenderer.sendSync('ELECTRON_BROWSER_MEMBER_CALL', metaId, member.name, wrapArgs(args))
//     return metaToValue(ret)
//   }
// } 
```

MarshallOfSound commented:
One has to think that this won't be the only function that needs to be explicitly bound to its correct context.  I wonder if things like `valueOf` and `isPrototypeOf` experience the same issue.

twolfson commented:
Ah, good thinking. After running those in the console, they seem to be fine:

```js
new Proxy(function x () {}, {}).valueOf()
// function anonymous()[[Handler]]: Object[[Target]]: function x()[[IsRevoked]]: false
new Proxy(function x () {}, {}).isPrototypeOf({})
// false
```

MarshallOfSound commented:
> // function anonymous()[[Handler]]: Object[[Target]]: function x()[[IsRevoked]]: false

Interesting, appears to be the `valueOf` call against a proxy object rather than the target 💭 TIL

kevinsawicki commented on spec/api-remote-spec.js:
```diff
@@ -0,0 +1,11 @@
+const electron = require('electron')
+
+describe('remote module', function () {
+  describe('remoteMemberFunction.toString()', function () {
```
Would you mind moving this to `api-ipc-spec.js` near the other `remote` specs? I know the name isn't super clear but that is where the other remote function specs are.

twolfson replied:
Sure thing, I'll move it over

twolfson commented:
Alright, I have relocated the test into `api-ipc-spec.js` as requested :+1:

kevinsawicki commented on lib/renderer/api/remote.js:
```diff
@@ -175,6 +175,7 @@ const proxyFunctionProperties = function (remoteMemberFunction, metaId, name) {
       return true
     },
     get: (target, property, receiver) => {
+      if (property === 'toString') return target.toString.bind(target)
```
Would you be okay adding a second guard here:

```js
if (property === 'toString' && typeof target.toString === 'function')
```

Just to be on the safe side in case they've overridden `toString` to be a property or something else.

twolfson replied:
I believe that's redundant as a proxy can only receive a function as a parameter. If this is a safety check for `Proxy's` definition supporting more than functions, then it might be better to use an `assert` instead so we can detect the new case via error reporting and implement its behavior accurately instead of trying to guess:

```js
if (property === 'toString') {
  // Sanity check that Proxy only supports functions
  assert(target.toString);

  // Return our original toString
  return target.toString.bind(target);
}
```

Thoughts?

MarshallOfSound replied:
@twolfson The case for `toString` being not the built-in function is quite easy to reproduce.

```js
var a = new Proxy({ toString: 'foo' }, { get: (a, b) => a[b] })
```

In this case `a.toString` returns the string "foo".  I think the safest check to do is this.

```js
if (property === 'toString' && target.constructor.prototype.toString === target.toString) {
  // Return bind
}
```

This ensures that not only is the toString property a function but it is the original built-in function that the object was created with.

twolfson replied:
Ah, I was recalling MDN docs from memory. I could have sworn it was function only, guess not =/

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy

@MarshallOfSound `target.constructor.prototype` could be another class with a custom `toString` as its property. It's best to go @kevinsawicki's route of verifying it's a method and letting JS handle the prototypical resolution

twolfson commented:
Alright, added the additional type check for `toString` being a function. I didn't add any tests as I can't think of any APIs that would have a non-function `toString` and felt like adding a one-off module would be overkill

kevinsawicki commented:
> Alright, added the additional type check for toString being a function. I didn't add any tests as I can't think of any APIs that would have a non-function toString and felt like adding a one-off module would be overkill

I added a test for a remote require that had a function with a `toString` property value that was not a function and had to made a tweak to your change to get it to pass.

@twolfson please take a look at my pushed changes and let me know if they look good to you.

twolfson commented:
Looks good to me 👍

On Mar 17, 2017 10:30 AM, "Kevin Sawicki" <notifications@github.com> wrote:

> Alright, added the additional type check for toString being a function. I
> didn't add any tests as I can't think of any APIs that would have a
> non-function toString and felt like adding a one-off module would be
> overkill
>
> I added a test for a remote require that had a function with a toString
> property value that was not a function and had to made a tweak to your
> change to get it to pass.
>
> @twolfson <https://github.com/twolfson> please take a look at my pushed
> changes and let me know if they look good to you.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/electron/electron/pull/8890#issuecomment-287420069>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA3FWAwik744JFj1pJJyHSEVCK_oG5wMks5rmsNLgaJpZM4MZFqd>
> .
>

kevinsawicki commented:
Thanks for this @twolfson 👍 🚀
